metronome/hotstuff/consensus/src/metronome/hotstuff/consensus/basic/ProtocolState.scala

```diff
@@ -0,0 +1,492 @@
+  *
+  * See https://arxiv.org/pdf/1803.05069.pdf
+  *
+  * ```
+  *
+  * PHASE                  LEADER                         REPLICA
+  *                           |                              |
+  *                           | <--- NewView(prepareQC) ---- |
+  * # Prepare --------------- |                              |
+  *             select highQC |                              |
+  *             create block  |                              |
+  *                           | ------ Prepare(block) -----> |
+  *                           |                              | check safety
+  *                           | <----- Vote(Prepare) ------- |
+  * # PreCommit ------------- |                              |
+  *                           | ------ Prepare Q.C. -------> |
+  *                           |                              | save as prepareQC
+  *                           | <----- Vote(PreCommit) ----- |
+  * # Commit ---------------- |                              |
+  *                           | ------ PreCommit Q.C. -----> |
+  *                           |                              | save as lockedQC
+  *                           | <----- Vote(Commit) -------- |
+  * # Decide ---------------- |                              |
+  *                           | ------ Commit Q.C. --------> |
+  *                           |                              | execute block
+  *                           | <--- NewView(prepareQC) ---- |
+  *                           |                              |
+  *
+  * ```
+  */
+case class ProtocolState[A <: Agreement: Block: Signing](
+    viewNumber: ViewNumber,
+    phase: Phase,
+    publicKey: A#PKey,
+    signingKey: A#SKey,
+    federation: Federation[A#PKey],
+    // Highest QC for which a replica voted Pre-Commit, because it received a Prepare Q.C. from the leader.
+    prepareQC: QuorumCertificate[A],
+    // Locked QC, for which a replica voted Commit, because it received a Pre-Commit Q.C. from leader.
+    lockedQC: QuorumCertificate[A],
+    // Hash of the block that was last decided upon.
+    lastExecutedBlockHash: A#Hash,
+    // Hash of the block the federation is currently voting on.
+    preparedBlockHash: A#Hash,
+    // Timeout for the view, so that it can be adjusted next time if necessary.
+    timeout: FiniteDuration,
+    // Votes gathered by the leader in this phase. They are guarenteed to be over the same content.
+    votes: Set[Message.Vote[A]],
+    // NewView messages gathered by the leader during the Prepare phase. Map so every sender can only give one.
+    newViews: Map[A#PKey, Message.NewView[A]]
+) {
+  import Message._
+  import Effect._
+  import Event._
+  import ProtocolState._
+  import ProtocolError._
+
+  val leader   = federation.leaderOf(viewNumber)
+  val isLeader = leader == publicKey
+
+  /** The leader has to collect `n-f` signatures into a Q.C. */
+  def quorumSize = federation.size - federation.maxFaulty
```

Comment:
Thank you for that explanation! I did it this way to follow the paper, but I also thought that it was a bit unintuitive that if we assume n=10 and f=1, then we have to get a quorum of 9, while f=3 requires (or allows) a smaller quorum size. Seeing the whole inequality makes more sense.

I thought this could be a separate PR, but on second thought I'll put it into this one, since `maxFaulty` can be set to 1 and then it doesn't make any sense the way it is.

Reply:
Ah, forgot that `maxFaulty` is not something you can set, it always uses the most pessimistic value, giving the largest quorum.

In that light, maybe a separate PR _is_ better. Do you see `f` being configurable?

Reply:
Created https://jira.iohk.io/browse/PM-3058

Reply:
I think we should fix `f = n/3-1`. That way both n-f and (n+f)/2 would be the same.

Reply:
Hm, that's kind of what's [happening now](https://github.com/input-output-hk/metronome/pull/5/files#diff-d56a67191a6bf58e5b4c7b05b74eed1db2fc992b462b4a77dda908974f645e64R12-R13), resulting both upper and lower thresholds be the same, albeit with `f = (n-1)/3` (so with 10 nodes it gives (10-1)/3=3, not 10/3-1=2). 

I thought the benefit of your formula would be that we could set an `f` which is lower than the theoretical maximum, and thus be content with a lower quorum size, e.g. set f=1 and get (10+1)/2+1=6 instead of 7.